src/3d/shaders/ssao_factor_render.frag

```diff
@@ -0,0 +1,76 @@
+#version 330
+
+uniform sampler2D depthTexture;
+
+uniform float farPlane;
+uniform float nearPlane;
+uniform mat4 origProjMatrix;      // Perspective projection matrix used for forward rendering
+
+uniform vec3 ssaoKernel[64];  // Random sample vectors in a unit sphere
+const int kernelSize = 64;
+uniform vec3 ssaoNoise[16];  // 4x4 random noise pattern
+
+uniform float	intensity;  // Amplification of shading
+uniform float	radius;     // Radius of neighborhood sphere (in world units)
+
+noperspective in vec3 vViewRay;   // Ray to far plane
+
+out vec4 fragColor;
+
+
+float linearizeDepth(float depth)
+{
+  float ndc = depth * 2.0 - 1.0;
+  return (2.0 * nearPlane * farPlane) / (farPlane + nearPlane - ndc * (farPlane - nearPlane));
+}
+
+vec3 rotate_x(vec3 vct, float angle)
+{
+  return vec3(vct.x*cos(angle)-vct.y*sin(angle), vct.x*sin(angle)+vct.y*cos(angle), vct.z);
+}
+vec3 rotate_y(vec3 vct, float angle)
+{
+  return vec3(vct.x*cos(angle)+vct.z*sin(angle), vct.y, -vct.x*sin(angle)+vct.z*cos(angle));
+}
+
+float ssao(vec3 originPos, vec3 noise)
+{
+    float occlusion = 0.0;
+    for (int i = 0; i < kernelSize; ++i)
```

Comment:
```
    //	project pixel position on screen:
    vec4 originPosScreen = origProjMatrix * vec4(originPos, 1.0);
    originPosScreen.xy = originPosScreen.xy / originPosScreen.w  * 0.5 + 0.5;

    //	get sample depth:
    float depth = texture(depthTexture, originPosScreen.xy).r;
    if ( depth == 1 )
      return 1.0f;
```

Reply:
This added code will avoid making the scene darker where nothing is rendered

Reply:
I think Nyall was concerned not only about the background (where depth==1), but also the foreground (terrain, buildings etc.), as highlighted in his screenshot, so I am afraid this is not going to help...

Reply:
Skipping the background definitely helps though, I believe this check should be included.
Also, lowering the default intensity to `0.5` instead of `1` gives a nicer looking result imho. The darkening is still there, but less... dark :)